hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java

```diff
@@ -378,6 +379,28 @@ private void startInstant() {
         this.conf.getString(FlinkOptions.TABLE_NAME), conf.getString(FlinkOptions.TABLE_TYPE));
   }
 
+  /**
+   * Get the valid instant time of last batch from bootstrap events
+   * Return Option.empty() to indicate the instant is invalid from last batch
+   */
+  protected Option<String> bootstrapInstantFromEventBuffer() {
+    ValidationUtils.checkArgument(Arrays.stream(eventBuffer).allMatch(evt -> evt != null && evt.isBootstrap()));
+    List<WriteMetadataEvent> events = Arrays.stream(eventBuffer).filter(evt -> !evt.getInstantTime().equals("")).collect(Collectors.toList());
+    String instant = events.stream().map(WriteMetadataEvent::getInstantTime).reduce((a, b) -> a.equals(b) ? a : "").orElse("");
+    // instant and parallelism should be unique
```

Comment:
The fix may be right but there are some thoughts about code engineering:

1. we should not put static config options in `WriteMetadataEvent` which is a dynamic metadata event from write task.

2. do we really need to care about the parallelism of last run ? Can we just merge the bootstrap events first before sending it to the coordinator ?

Reply:
We have to persist the parallelism somewhere because it is used to validate if all flushes in the last run complete. And it may change across runs. So I choose the most straightforward way to store it (i.e., in metadata state). Another option is to let user to pass in parallelism of the last run through configs, which is not user-friendly as users need to understand the logic behind it.

Merging bootstrap events before sending it a good point! Will improve it and seems we could remove the transient variable `mergeCount` based on it.

Reply:
> We have to persist the parallelism somewhere because it is used to validate if all flushes in the last run complete. And it may change across runs

Did you notice that we send an empty bootstrap event even if there is nothing in the metadata state ?

Reply:
Yes, and those empty event will be excluded as the instant time is empty.

Lets say we increase `write.task` from 4 -> 8. During restart, we will receive 8 metadata event (with 4 empty bootstrap event). In order to do know if the remaining non-empty metadata events are the full collection of events from last run, we have to get the parallelism of the last run (i.e., 4) rather than checking using the new parallelism (i.e., 8)

Reply:
The operator state backend would spread the events evenly within the write tasks, and i want to say that the empty bootstrap event is also valid for coordinator validation.

But here is a problem, when we decrease the parallelism, one write task may have several bootstrap events, we should merge these events first before sending them to coordinator, or the coordinator may commit eagerly before it receives all the bootstrap events.

Reply:
Yes, so I add `eventNumOfTask`/`mergeCount` variables and change the eager commit behavior. Following your 'merge before send' suggestion, the `mergeCount` is no longer needed.

Reply:
Yes, merge events before sending to coordinator makes the logic much simpler.